src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureKeyVaultReference/AzureKeyVaultSecretProvider.cs

```diff
@@ -48,12 +43,25 @@ public async Task<string> GetSecretValue(Uri secretUri, CancellationToken cancel
 
             if (client != null)
             {
-                KeyVaultSecret secret = await client.GetSecretAsync(secretName, secretVersion, cancellationToken).ConfigureAwait(false);
-                secretValue = secret?.Value;
+                KeyVaultSecret secret;
+
+                if (_cachedKeyVaultSecrets.TryGetValue(key, out CachedKeyVaultSecret cachedSecret) &&
+                    (!cachedSecret.RefreshAt.HasValue || DateTimeOffset.UtcNow < cachedSecret.RefreshAt.Value))
+                {
+                    secretValue = cachedSecret.SecretValue;
+                }
+                else
+                {
+                    // We dont have a cached secret value or the cached value has expired.
+                    // Get the secret from Key Vault and update the cache.
+                    secret = await client.GetSecretAsync(secretName, secretVersion, cancellationToken).ConfigureAwait(false);
+                    secretValue = secret?.Value;
+                    SetSecretInCache(key, secretValue);
+                }
             }
-            else if (_secretResolver != null)
+            else if (_keyVaultOptions.SecretResolver != null)
             {
-                secretValue = await _secretResolver(secretUri).ConfigureAwait(false);
+                secretValue = await _keyVaultOptions.SecretResolver(secretUri).ConfigureAwait(false);
```

Comment:
>_keyVaultOptions.SecretResolver [](http://example.com/codeflow?start=36&length=31)

How come we're not caching secrets returned by secret resolver?

Reply:
We don't know what users have implemented in SecretResolver. They may want the callback to run every time instead of caching the value. Also, if users are defining a secret resolver, I think it's safe to assume that they are responsible for managing the custom implementation and optimizing the callback themselves.

Reply:
But in this case we don't give them the option to rotate secrets/certificates

Reply:
That's correct. This callback is not meant for getting secrets/certs in the first place. It was introduced as a way to avoid resolving secrets from Key Vault by assigning a dummy value to key vault references. I don't think there should be an expectation to provide the same caching behavior for this custom implementation.

Reply:
I'd say that's half true. Yes we designed the API because of a scenario where users wanted to skip resolving certain secrets. However, the API we designed is one that allows custom resolution of secrets instead of KeyVaultClient. This means that it is fine to use for custom resolution of secrets. 

Caching + rotation is a value add a layer above the resolution of secrets. I don't think it should depend on what client/resolver we use to get a secret.

Reply:
That makes sense. The only thing is, users don't have an option to opt out of caching+rotation for the custom resolved secrets (that is if they want to use rotation at all). If they use the `SetSecretRefreshInterval(refreshInterval)` API, everything under the custom callback will also be cached. 
But maybe that's not such a big issue. I can't think of a use case where someone would want this. So I'll update the code to cache custom secrets too.